# Features selection

* This notebook contains code for selection of important features from the dataset.
* The important features are selected and filtered from the dataset and stored in a new file.
* The definitions of the selected features are provided below for reference.

## Importing modules

In [1]:
import pandas as pd

## Loading data from CSV file

In [9]:
df = pd.read_csv('genre-data-clean.csv')
df.head()

Unnamed: 0  danceability  energy  key  loudness  mode  speechiness  \
0           0         0.507  0.7100    5    -5.020     1       0.0291   
1           1         0.496  0.3780   11    -7.878     1       0.0336   
2           2         0.760  0.7650    0    -9.319     1       0.0352   
3           3         0.607  0.1960    1   -19.705     1       0.0595   
4           4         0.497  0.0919    3   -15.685     1       0.0704   

   acousticness  instrumentalness  liveness  ...    tempo            type  \
0        0.0975          0.000011     0.170  ...  134.647  audio_features   
1        0.8980          0.000001     0.109  ...  144.092  audio_features   
2        0.2890          0.000008     0.233  ...  107.602  audio_features   
3        0.9040          0.727000     0.111  ...  111.031  audio_features   
4        0.9680          0.002140     0.133  ...  206.431  audio_features   

                       id                                   uri  \
0  14J3PO0VnhtcRa31r7Aj1L  spotify:track:14J3PO0VnhtcRa31r7Aj1L   
1  2SZDu4esDLUISRnpYQE17b  spotify:track:2SZDu4esDLUISRnpYQE17b   
2  5E2EpolcGmcbYpsv0z6jAZ  spotify:track:5E2EpolcGmcbYpsv0z6jAZ   
3  5RVEXH7OR9rcRz7eeviOzc  spotify:track:5RVEXH7OR9rcRz7eeviOzc   
4  7CNaYAdLyi86kofGafReiT  spotify:track:7CNaYAdLyi86kofGafReiT   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/14J3PO0Vnhtc...   
1  https://api.spotify.com/v1/tracks/2SZDu4esDLUI...   
2  https://api.spotify.com/v1/tracks/5E2EpolcGmcb...   
3  https://api.spotify.com/v1/tracks/5RVEXH7OR9rc...   
4  https://api.spotify.com/v1/tracks/7CNaYAdLyi86...   

                                        analysis_url duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/14J3...      273573   
1  https://api.spotify.com/v1/audio-analysis/2SZD...      277667   
2  https://api.spotify.com/v1/audio-analysis/5E2E...      226933   
3  https://api.spotify.com/v1/audio-analysis/5RVE...      223948   
4  https://api.spotify.com/v1/audio-analysis/7CNa...      139693   

   time_signature  year     genre  
0               4  1999  acoustic  
1               3  1999  acoustic  
2               4  1999  acoustic  
3               4  1999  acoustic  
4               4  1999  acoustic  

[5 rows x 21 columns]

## Removing unnecessary features

There are 21 columns in the dataset. Below is a list of all the columns.

In [10]:
df.columns

Index(['Unnamed: 0', 'danceability', 'energy', 'key', 'loudness', 'mode',
       'speechiness', 'acousticness', 'instrumentalness', 'liveness',
       'valence', 'tempo', 'type', 'id', 'uri', 'track_href', 'analysis_url',
       'duration_ms', 'time_signature', 'year', 'genre'],
      dtype='object')

It can be identified that the following columns have no significance in training a machine learning mode:

1. `Unnamed: 0`
2. `type`
3. `id`
4. `uri`
5. `track_href`
6. `analysis_url`
7. `year`

Therefore, these columns are dropped from the dataset.

In [11]:
df_new = df.drop(
    ['Unnamed: 0', 'type', 'id', 'uri', 'track_href', 'analysis_url', 'year'], 
    axis=1, 
) 
df_new.head()

danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.507  0.7100    5    -5.020     1       0.0291        0.0975   
1         0.496  0.3780   11    -7.878     1       0.0336        0.8980   
2         0.760  0.7650    0    -9.319     1       0.0352        0.2890   
3         0.607  0.1960    1   -19.705     1       0.0595        0.9040   
4         0.497  0.0919    3   -15.685     1       0.0704        0.9680   

   instrumentalness  liveness  valence    tempo  duration_ms  time_signature  \
0          0.000011     0.170    0.450  134.647       273573               4   
1          0.000001     0.109    0.335  144.092       277667               3   
2          0.000008     0.233    0.868  107.602       226933               4   
3          0.727000     0.111    0.187  111.031       223948               4   
4          0.002140     0.133    0.548  206.431       139693               4   

      genre  
0  acoustic  
1  acoustic  
2  acoustic  
3  acoustic  
4  acoustic

## Writing filtered dataset to file

In [13]:
df_new.to_csv('music_genre_dataset.csv', index=False)

## Definitions of Various Audio Features

### Acousticness

A confidence measure from 0.0 to 1.0 of whether the track is acoustic. 1.0 represents high confidence the track is acoustic.

### Danceability

Danceability describes how suitable a track is for dancing based on a combination of musical elements including tempo, rhythm stability, beat strength, and overall regularity. A value of 0.0 is least danceable and 1.0 is most danceable.

### Energy

Energy is a measure from 0.0 to 1.0 and represents a perceptual measure of intensity and activity. Typically, energetic tracks feel fast, loud, and noisy. For example, metal genre has high energy, while a Bach prelude scores low on the scale. Perceptual features contributing to this attribute include dynamic range, perceived loudness, timbre, onset rate, and general entropy.

### Instrumentalness

Predicts whether a track contains no vocals. "Ooh" and "aah" sounds are treated as instrumental in this context. Rap or spoken word tracks are clearly "vocal". The closer the instrumentalness value is to 1.0, the greater likelihood the track contains no vocal content. Values above 0.5 are intended to represent instrumental tracks, but confidence is higher as the value approaches 1.0.

### Key

The key the track is in. Integers map to pitches using standard Pitch Class notation. E.g. 0 = C, 1 = C♯/D♭, 2 = D, and so on. If no key was detected, the value is -1.

### Liveliness

Detects the presence of an audience in the recording. Higher liveness values represent an increased probability that the track was performed live. A value above 0.8 provides strong likelihood that the track is live.

### Loudness

The overall loudness of a track in decibels (dB). Loudness values are averaged across the entire track and are useful for comparing relative loudness of tracks. Loudness is the quality of a sound that is the primary psychological correlate of physical strength (amplitude). Values typically range between -60 and 0 db.

### Mode

Mode indicates the modality (major or minor) of a track, the type of scale from which its melodic content is derived. Major is represented by 1 and minor is 0.

### Speechiness

Speechiness detects the presence of spoken words in a track. The more exclusively speech-like the recording (e.g. talk show, audio book, poetry), the closer to 1.0 the attribute value. Values above 0.66 describe tracks that are probably made entirely of spoken words. Values between 0.33 and 0.66 describe tracks that may contain both music and speech, either in sections or layered, including such cases as rap music. Values below 0.33 most likely represent music and other non-speech-like tracks.

### Tempo

The overall estimated tempo of a track in beats per minute (BPM). In musical terminology, tempo is the speed or pace of a given piece and derives directly from the average beat duration.

### Time Signature

An estimated time signature. The time signature (meter) is a notational convention to specify how many beats are in each bar (or measure). The time signature ranges from 3 to 7 indicating time signatures of "3/4", to "7/4".

### Valence

A measure from 0.0 to 1.0 describing the musical positiveness conveyed by a track. Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry).

## Note

Features like `mode`, `speechiness`, `liveliness` do not have much impact on the genre of music. However, these features can be used for further statistical analysis.